### ISSUE TITLE
[FEATURE REQUEST] Smart handling of S.M.A.R.T.

### ISSUE THREAD

**Issue by danielvandenberg95** [ngandrass/truenas-spindown-timer#25]
Like some others, I needed to disable smart on my drives to make this script useful. Keeping smart on spins the disks up after max 5 minutes of inactivity.

I don't like smart being off.

A: Would it be possible to turn off smart when spinning a disk down, and turn it on again when it's detected that the disk is running again?
B: Could this be added to the readme, that SMART (sometimes?) needs to be turned off? I found out about it by reading through closed issues...

**Comment by ngandrass** [comment]
> A: Would it be possible to turn off smart when spinning a disk down, and turn it on again when it's detected that the disk is running again?

I tried to investigate this issue here: https://github.com/ngandrass/truenas-spindown-timer/issues/14#issuecomment-1491809880

The problem seems to be, that the TrueNAS S.M.A.R.T. config generator always enables temperature monitoring (`-W`) and `smartd` ignores the current disk power mode during this check.


> B: Could this be added to the readme, that SMART (sometimes?) needs to be turned off? I found out about it by reading through closed issues...

I'd rather like to see the bug within TrueNAS fixed than suggesting to disable S.M.A.R.T., as the whole point of this script is to be able to spin down drives **while the S.M.A.R.T. service is active**.

I could file an issue and try to provide a potential fix. However, I'll be on vacation the next three weeks and probably won't find the time required to do so ... I'll have to look into it when I'm back, sorry :(

**Comment by Dark-Noir** [comment]
What about manual editing the /etc/smartd.conf ?

It is possible (at least with other debian systems) to configure a standby/spindown check.
So that they not spin up (only f.e. when too many checks are missed, configurable).

**Comment by ngandrass** [comment]
That seems possible as a quick and dirty workaround. However, keep in mind that changes within the TrueNAS config/GUI can cause `/etc/smartd.conf` to be fully overwritten, leaving you with constantly spinning disks at some point.

**Comment by Dark-Noir** [comment]
I configured the S.M.A.R.T service in the GUI like this:
![Clipboard01](https://github.com/ngandrass/truenas-spindown-timer/assets/8070059/89f95b0f-ac7d-4544-94b3-41ef719502e6)
This modifies the `/etc/smartd.conf` like this:
```
/dev/sdb -a -d removable -n standby -W 5,45,50 -m root -M exec /usr/local/libexec/smart_alert.py
/dev/sdc -a -d removable -n standby -W 5,45,50 -m root -M exec /usr/local/libexec/smart_alert.py
/dev/sde -a -d removable -n standby -W 5,45,50 -m root -M exec /usr/local/libexec/smart_alert.py
/dev/sdf -a -d removable -n standby -W 5,45,50 -m root -M exec /usr/local/libexec/smart_alert.py
/dev/sdg -a -d removable -n standby -W 5,45,50 -m root -M exec /usr/local/libexec/smart_alert.py
/dev/sdh -a -d removable -n standby -W 5,45,50 -m root -M exec /usr/local/libexec/smart_alert.py
``` 
But the disks still spin up after some time.
The GUI misses some important settings, so i configured my own `/etc/smartd.conf`:
```
DEVICESCAN -a -u -p -n standby,24,q \
-I 194 \
-I 190 \
-W 5,45,50 \
-i 9 \
-R 5! \
-o on -S on -s (S/../.././02|L/../01/./04) \
-m root \
-M exec /usr/local/libexec/smart_alert.py
``` 
Doc: https://linux.die.net/man/5/smartd.conf

But i think it gets reseted after reboot, so a small script or cronjob is necessary.

**Comment by danielvandenberg95** [comment]
> 

Awesome, did not see that issue as it's title is not exactly descriptive of this issue. For now I'll leave SMART turned off. Enjoy your holidays!

**Comment by ngandrass** [comment]
Yeah, #14 derailed a bit... I changed the title so that it can be found more easily. I'll close this here for now and will tackle the smartd config problems once I'm back. Hopefully I'll find some time during the holidays to debug this further :)

**Comment by Sophist-UK** [comment]
I tried using the TrueNas Scale standard HDD power off and it worked but SMART turned the drives on again.

So I found this script and the ReadMe says:

>Periodic reads of S.M.A.R.T. data performed by the smartctl service are excluded. This allows users to have S.M.A.R.T. reporting enabled while being able to automatically spin down disks. The script moreover is immune to the periodic disk temperature reads in newer versions of TrueNAS.

So I just implemented this script following the instructions in the ReadMe.

But now I am reading this Issue and wondering whether SMART stuff is excluded and if it isn't and I need to turn SMART off to get spindowns to work as expected, then what exactly is the difference between this script and the standard TrueNas functionality?

**Comment by Sophist-UK** [comment]
I just set the SMART Service to use `-n standby`, so when I get the chance I will need to see whether the standard TrueNas standby functionality and / or this script keeps the disks off now this has been changed.

**Comment by brendanlees** [comment]
I might be wrong but even if the config was successfully changed by editing smartd.conf (I could not get this to work), auto SMART tests will only run in TrueNAS if the disk(s) mode is set to 'never' (not 'standby')?

This would seem to indicate that a default would be to maintain 'standby' to keep the drives spun down, then preemptively changing the mode to 'never' before a scheduled test, and then reverting it back to 'standby' after so disks can spin down again. 

Not simple, but all of this could theoretically be planned with cron jobs, and is a better theoretical solution than having no tests at all.

Lastly the other issue I found recently is that when disk mode is in 'never', if spindowntimer is running it will override and spin down drives while a test is running. So a running instance of the spinddowntimer script itself would probably need to be disabled for test duration as well 🤷‍♂️

**Comment by Sophist-UK** [comment]
I can see that we wouldn't want to issue SMART or Temp. sense commands to a disk which is spundown if this would cause it to spin up again. So you would want either Linux (or Truenas) to prevent this.

I can also see that we wouldn't want to issue these commands to a disk if they would be considered proper I/O and reset the idle timer. I am not sure whether the standard spindown functionality is in hardware or software, but looking at the [man page for `hdparm -S`](https://man7.org/linux/man-pages/man8/hdparm.8.html) suggests that since only some drives support this in hardware it is most likely done in software - if it is indeed done in software, it should be possible for the software to exclude SMART/Temp. sense I/Os from the counts.

BUT, the idea that either Linux or Truenas would avoid doing SMART/Temp. reads at all because you have set a spindown time is ridiculous. These readings are at least **very important** in a NAS (and some would consider them essential) - excessive temperatures are a **major** cause of hard drive failures both because they can indicate a drive bearing issue and because the temperatures themselves can cause degradation, so Linux / Truenas should do their absolute best to get these readings providing that they don't either wake a spundown disk or stop a disk going into standby. At present this does NOT seem to be the case.

There are several statements in the Linux/Truenas documentation about why SMART/Temp. sensing is not done, and I am genuinely uncertain whether these are completely truthful or not.
